node/core/dispute-coordinator/src/import.rs

```diff
@@ -0,0 +1,533 @@
+		Self {
+			votes,
+			own_vote: OwnVoteState::NoVote,
+			concluded_invalid: false,
+			concluded_valid: false,
+			is_confirmed: false,
+			is_disputed: false,
+		}
+	}
+
+	/// Create a new `VoteState` from already existing votes.
+	pub fn new<'a>(votes: CandidateVotes, env: &CandidateEnvironment<'a>) -> Self {
+		let own_vote = OwnVoteState::new(&votes, env);
+
+		let n_validators = env.validators().len();
+
+		let supermajority_threshold =
+			polkadot_primitives::v2::supermajority_threshold(n_validators);
+
+		let concluded_invalid = votes.invalid.len() >= supermajority_threshold;
+		let concluded_valid = votes.valid.len() >= supermajority_threshold;
+
+		// We have a dispute, if we have votes on both sides:
+		let is_disputed = !votes.invalid.is_empty() && !votes.valid.is_empty();
+
+		let byzantine_threshold = polkadot_primitives::v2::byzantine_threshold(n_validators);
+		let is_confirmed = votes.voted_indices().len() > byzantine_threshold && is_disputed;
+
+		Self { votes, own_vote, concluded_invalid, concluded_valid, is_confirmed, is_disputed }
+	}
+
+	/// Import fresh statements.
+	///
+	/// Result will be a new state plus information about things that changed due to the import.
+	pub fn import_statements(
+		self,
+		env: &CandidateEnvironment,
+		statements: Vec<(SignedDisputeStatement, ValidatorIndex)>,
+	) -> ImportResult {
+		let (mut votes, old_state) = self.into_old_state();
+
+		let mut new_invalid_voters = Vec::new();
+		let mut imported_invalid_votes = 0;
+		let mut imported_valid_votes = 0;
+
+		let expected_candidate_hash = votes.candidate_receipt.hash();
+
+		for (statement, val_index) in statements {
+			if env
+				.validators()
+				.get(val_index.0 as usize)
+				.map_or(true, |v| v != statement.validator_public())
+			{
+				gum::error!(
+				target: LOG_TARGET,
+				?val_index,
+				session= ?env.session_index,
+				claimed_key = ?statement.validator_public(),
+				"Validator index doesn't match claimed key",
+				);
+
+				continue
+			}
+			if statement.candidate_hash() != &expected_candidate_hash {
+				gum::error!(
+				target: LOG_TARGET,
+				?val_index,
+				session= ?env.session_index,
+				given_candidate_hash = ?statement.candidate_hash(),
+				?expected_candidate_hash,
+				"Vote is for unexpected candidate!",
+				);
+				continue
+			}
+			if statement.session_index() != env.session_index() {
+				gum::error!(
+				target: LOG_TARGET,
+				?val_index,
+				session= ?env.session_index,
+				given_candidate_hash = ?statement.candidate_hash(),
+				?expected_candidate_hash,
+				"Vote is for unexpected session!",
+				);
+				continue
+			}
+
+			match statement.statement() {
+				DisputeStatement::Valid(valid_kind) => {
+					let fresh = insert_into_statements(
+						&mut votes.valid,
+						*valid_kind,
+						val_index,
+						statement.into_validator_signature(),
+					);
+
+					if fresh {
+						imported_valid_votes += 1;
+					}
```

Comment:
```suggestion
					imported_valid_votes += insert_into_statements(
						&mut votes.valid,
						*valid_kind,
						val_index,
						statement.into_validator_signature(),
					) as u32;
```

Reply:
I'd say the former is better, couple of lines for readability